File: electives/apm/lab/lista-1/ex-2/java/app/src/main/java/com/example/ordnung/InvertibleMap.java
package com.example.ordnung;

import java.util.HashMap;
import java.util.function.BiConsumer;

public class InvertibleMap<T, S> {
    private HashMap<T, S> primary = new HashMap<>();
    private HashMap<S, T> inverted = new HashMap<>();

    public InvertibleMap(HashMap<T, S> first) {
        first.forEach(
            (t, s) -> {
                this.primary.put(t, s);
                this.inverted.put(s, t);
            }
        );
    }

    public S getObj(T key) {
        return this.primary.get(key);
    }

    public T getKey(S obj) {
        return this.inverted.get(obj);
    }

    public void forEach(BiConsumer<T, S> action) {
        this.primary.forEach(action);
    }

    public int size() {
        return this.primary.size();
    }

}

File: electives/apm/lab/lista-2/ex-1/TicTacToe/app/src/main/java/com/example/tictactoe/gamelogic/InvertibleMap.kt
package com.example.tictactoe.gamelogic

import java.util.*

class InvertibleMap<T, S> {

    private val primary = HashMap<T, S>()
    private val inverted = HashMap<S, T>()
    fun getObj(key: T): S? {
        return primary[key]
    }

    fun getObjs(): List<S> {
        return this.primary.values.toList()
    }

    fun getKey(obj: S): T? {
        return inverted[obj]
    }

    fun forEach(action: (Map.Entry<T, S>) -> Unit) {
        primary.forEach(action)
    }

    fun size(): Int {
        return primary.size
    }

    constructor(first: HashMap<T, S>) {
        first.forEach { (t: T, s: S) ->
            primary[t] = s
            inverted[s] = t
        }
    }
}
